controlplane/kubeadm/internal/proxy/dial.go

```diff
@@ -92,7 +96,17 @@ func (d *Dialer) DialContext(_ context.Context, _ string, addr string) (net.Conn
 		Name(addr).
 		SubResource("portforward")
 
-	dialer := spdy.NewDialer(d.upgrader, &http.Client{Transport: d.proxyTransport}, "POST", req.URL())
+	spdyDialer := spdy.NewDialer(d.upgrader, &http.Client{Transport: d.proxyTransport}, "POST", req.URL())
+
+	websocketDialer, err := portforward.NewSPDYOverWebsocketDialer(req.URL(), d.restConfig)
```

Comment:
Do we know if this only works with certain Kubernetes apiserver versions?

Reply:
Without diving to deep into the upstream k/k PR it looks to me like there is a server-side feature-gate to make this possible (PortForwardWebsockets) and it's only enabled per default with 1.31?
Is this correct?

xref:
* https://github.com/kubernetes/kubernetes/pull/125528
* https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/


If yes, this means for Kubernetes < 1.31 we are now with every single reconcile trying the websocketDialer but it won't work. Sounds like a lot of unnecessary "work" (not sure what it's exactly doing, requests?). We should probably try to avoid this

Reply:
+1
Wondering if we should have two dials method and invoke them depending on the k8s version, or we should have a cache (might be a TTL cache) where we track IP that do not support web sockets

Reply:
We maybe should configure the fallback depending on the k8s version. 

< 1.31 : spdy first, websocket second
>= 1.31 : websocket first, spdy second

?

Reply:
That would be the simple option, it just doesn't take care of the case where someone disables the feature gate in 1.31.

Do we have an idea of how high the impact is of falling back on every single reconcile?

Reply:
Ah wait. We're talking about KCP, so KCP would actually now if someone enables/disables the feature gate (and not just the default is used)